### ISSUE TITLE
Handle /usr/lib/locale and /usr/share/locale

### ISSUE THREAD

**Issue by probonopd** [VHSgunzo/sharun#3]
Some Linux distributions such as NixOS and Chimera Linux don't have `/usr/lib/locale` and `/usr/share/locale`, so at least for glibc-based applications these need to be bundled.

**Comment by Samueru-sama** [comment]
Do we always need to deploy the locales btw? For example I found out we don't always need to deploy `gconv` from deploying glibc, something that go-appimage will always do as far as I know.

**Comment by probonopd** [comment]
At least PrusaSlicer is giving me

![image](https://github.com/user-attachments/assets/4de2da08-0e3c-49d4-9373-6b627993ec1c)

on NixOS...

**Comment by Samueru-sama** [comment]
> on NixOS...

I assume that the error goes away when you copy `/usr/lib/locale` and `/usr/share/locale` over right? or is there a env variable that also needs to be set? sharun already sets `XDG_DATA_DIRS` to include the `share` dir but I'm not sure if there is a variable for `/usr/lib/locale`.

**Comment by probonopd** [comment]
Don't quite know what is going on but as soon as I set the `LOCPATH` environment variable I get the error...

**Comment by Samueru-sama** [comment]
Just reproduced the issue with GIMP, setting `LOCPATH` to even the location of the host locales breaks them 🤔

**Comment by probonopd** [comment]
On Debian, indeed `LOCPATH=/usr/lib/locale:/usr/share/locale ./AppRun` fails, and without setting LOCPATH it runs (and I think I just used the Debian default paths in this test). Strange. Need to find out whether this is caused by `sharun`.

**Comment by Samueru-sama** [comment]
> On Debian, indeed `LOCPATH=/usr/lib/locale:/usr/share/locale ./AppRun` fails, and without setting LOCPATH it runs (and I think I just used the Debian default paths in this test). Strange. Need to find out whether this is caused by `sharun`.

I tested with the arch gimp package and it breaks as well. 

![image](https://github.com/user-attachments/assets/fdd4775e-faaa-45f0-9c36-8a68acdf7149)

**Comment by probonopd** [comment]
I wonder what happens if you do the same with a distribution-installed gimp.

**Comment by Samueru-sama** [comment]
> I wonder what happens if you do the same with a distribution-installed gimp.

Isn't the gimp I just tested a distribution-installed gimp?

**Comment by probonopd** [comment]
Sorry, didn't catch this. So either we are using `LOCPATH` wrong or or it is entirely broken?

**Comment by Samueru-sama** [comment]
> Sorry, didn't catch this. So either we are using `LOCPATH` wrong or or it is entirely broken?

https://www.gnu.org/software/libc/manual/html_node/Locale-Names.html

According to that the default for `LOCPATH` is `/usr/share/locale` but it breaks if I set it to that. 

![image](https://github.com/user-attachments/assets/9699874a-8607-4e4a-bcf9-35ac91fc443a)

**Comment by Samueru-sama** [comment]
Qt apps will also complain about this, so hopefully I'm doing something wrong:

![image](https://github.com/user-attachments/assets/a8a02af9-2afa-4d77-9180-4ed7a4a2b8ce)

**Comment by probonopd** [comment]
According to https://www.man7.org/linux/man-pages/man7/locale.7.html, 

> LOCPATH
              A list of pathnames, separated by colons (':'), that
              should be used to find locale data.  If this variable is
              set, only the individual compiled locale data files from
              LOCPATH and the system default locale data path are used;
              **any available locale archives are not used** (see
              [localedef(1)](https://www.man7.org/linux/man-pages/man1/localedef.1.html)).  The individual compiled locale data files
              are searched for under subdirectories which depend on the
              currently used locale.  For example, when en_GB.UTF-8 is
              used for a category, the following subdirectories are
              searched for, in this order: en_GB.UTF-8, en_GB.utf8,
              en_GB, en.UTF-8, en.utf8, and en.

I wonder whether this is the reason.

**Comment by Samueru-sama** [comment]
haha, I was checking strings on all the libs to see which one contained `/usr/lib/locale-archive` to see if patching it works and just noticed you recent commit.

**Comment by probonopd** [comment]
Another option:
At least on my Ubuntu, /usr/lib/locale/ did not contain many files. Installing `sudo apt install locales-all` populated them. Now `LOCPATH` is no longer broken. Maybe we need to bundle the contents of that package if the libarchive-locale patching doesn't work out.

**Comment by probonopd** [comment]
`sed -i -e 's|/usr/lib/locale|././/lib/locale|g' ./shared/lib/libc.so.6` works, the bundled lib/locale-archive gets used.

**Comment by Samueru-sama** [comment]
> then bundling the contents of the `locales-all` Debian package would likely also work.

Noted, and yeah I've noticed I was locale errors when running other appimages on alpine but they didn't stop the app from working, this is really needed because changing the working directory is an issue with some apps. 

>with --appimage-extract for now

Doesn't it work if you set `FUSERMOUNT_PROG="$(command -v fusermount)"`?

**Comment by probonopd** [comment]
`FUSERMOUNT_PROG=$(which fusermount) ./My.AppImage` works nicely. So this is something we definitely need to fix in the AppImage type2-runtime. Apparently it doesn't search for `fusermount` on the `PATH` properly yet.

* https://github.com/AppImage/type2-runtime/issues/31

**Comment by VHSgunzo** [comment]
It will work if you just use uruntime. And it is also likely to help reduce the size of the appimage and speed up its work

**Comment by VHSgunzo** [comment]
@probonopd 
```
./uruntime-appimage-dwarfs-x86_64 --appimage-mkdwarfs -f --set-owner 0 --set-group 0 --no-history --no-create-timestamp --header uruntime-appimage-dwarfs-x86_64 -i squashfs-root -o PrusaSlicer-2.9.0-alpha1-x86_64.dwfs.AppImage --compression zstd:level=22 -S20 -B9
```
![image](https://github.com/user-attachments/assets/48f02b50-74b4-4a77-a583-c2cdf7efaf34)

**Comment by probonopd** [comment]
Very promising @VHSgunzo. Will come to uruntime later. Might consider it as the default if it turns out as great as it sounds.

**Comment by probonopd** [comment]
So, for my personal test case I found solutions for all of the above. But it is a lot of manual work, Is it in the scope of this project to provide automation for this? Thanks.

**Comment by Samueru-sama** [comment]
[Looks like all locales have to be bundled siince the patch on libc.so isn't enough. ](https://github.com/prusa3d/PrusaSlicer/issues/13675#issuecomment-2512000392)

**Comment by probonopd** [comment]
@Samueru-sama thanks for the pointer. Regarding PrusaSlicer issue 13653, do you have more insight into what exactly needs to be changed?

**Comment by Samueru-sama** [comment]
> @Samueru-sama thanks for the pointer. Regarding PrusaSlicer issue 13653, do you have more insight into what exactly needs to be changed?

The user that had the locale issue is from Romania, so I guess the locale that's used there isn't available in the `locale-archive` of the AppImage.

**Comment by Samueru-sama** [comment]
This issue is almost solved now with https://github.com/VHSgunzo/sharun/pull/92

We don't really need to bundle all the locales, just let the application use the ones of the host. This means all that is left is adding a check to make sure the locales of the host work before falling back to the bundled ones.

[I do that with this binary](https://github.com/pkgforge-dev/Anylinux-AppImages/blob/d23caaaac1313662413d4d4ad330daca83ffaad1/useful-tools/quick-sharun.sh#L1146-L1147) named `locale-check` and this logic in the `AppRun`: 

```shell
# check host locale works
if command -v locale-check 1>/dev/null && ! locale-check; then
	# use our locales
	__lcdir="$APPDIR"/shared/lib/locale
	[ ! -d "$__lcdir" ] || export LOCPATH="$__lcdir"
	# fallback to bundled C.UTF-8
	locale-check || export LC_ALL=C.UTF-8
	# fallback to bare C locale if this still fails somehow
	locale-check || export LC_ALL=C
fi
```

Basically if `setlocale` fails, then: 

* set `LOCPATH` to the bundled locales and run `setlocale` again. 

* If that fails, then `LC_ALL=C.UTF-8` (which is bundled) and run `setlocale` again.

* If that still fails, then `export LC_ALL=C`.

Ideally this logic should be in sharun, but I have no idea how to run `setlocale` in rust 😅
